### PR TITLE
Update emscripten docker image to be based on emscripten 2.0.12 and boost 1.75.0.

### LINKED ISSUE #10747: Bump emscripten in the build system
We agreed a while to periodically upgrade emscripten to avoid irregular large bumps.

Last we upgraded to emscripten 1.39.15 in solidity 0.6.9. Since then there has been 5 releases in 1.39x/1.40.x, and 13 releases in a new major version 2.x.y.

It would be useful to check the changelog and upgrade: https://github.com/emscripten-core/emscripten/blob/master/ChangeLog.md

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes https://github.com/ethereum/solidity/issues/10747

### CODE REVIEW

File: scripts/docker/buildpack-deps/Dockerfile.emscripten [hunk 1]
```diff
@@ -40,28 +40,27 @@ RUN set -ex; \
...
 		-DZ3_SINGLE_THREADED=ON \
 		-DCMAKE_CXX_FLAGS="-s DISABLE_EXCEPTION_CATCHING=0" \
 		..; \
-	make; make install; \
+	make ; make install; \
 	rm -r /usr/src/z3; \
 	cd /usr/src; \
-	wget -q 'https://dl.bintray.com/boostorg/release/1.73.0/source/boost_1_73_0.tar.bz2' -O boost.tar.bz2; \
-	test "$(sha256sum boost.tar.bz2)" = "4eb3b8d442b426dc35346235c8733b5ae35ba431690e38c6a8263dce9fcbb402  boost.tar.bz2"; \
+	wget -q 'https://dl.bintray.com/boostorg/release/1.75.0/source/boost_1_75_0.tar.bz2' -O boost.tar.bz2; \
```
Comment (axic) [inline scripts/docker/buildpack-deps/Dockerfile.emscripten:54]: Why is the boost upgrade needed?

Comment (ekpyron) [reply scripts/docker/buildpack-deps/Dockerfile.emscripten:54]: Because there's a new boost version ;-).

Comment (ekpyron) [reply scripts/docker/buildpack-deps/Dockerfile.emscripten:54]: We've done it at the same time in the past as well - there's little reason to expect issues with it and it halves the amount of work updating all this is.

**Comment by ekpyron** [comment]
Argh... I forgot that the docker images try a build and this immediately needs the build system updates in the same PR... that's not a good setup actually, because that'll make CI fail in this PR...

File: cmake/EthCompilerSettings.cmake [hunk 2]
```diff
@@ -150,7 +150,8 @@ if (("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU") OR ("${CMAKE_CXX_COMPILER_ID}" MA
 			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s WASM_ASYNC_COMPILATION=0")
 			# Output a single js file with the wasm binary embedded as base64 string.
 			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s SINGLE_FILE=1")
-
+			# Allow new functions to be added to the wasm module via addFunction.
+			set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -s ALLOW_TABLE_GROWTH=1")
```
Comment (ekpyron) [inline cmake/EthCompilerSettings.cmake:154]: This replaces ``RESERVED_FUNCTION_POINTERS``.

File: scripts/ci/build_emscripten.sh [hunk 3]
```diff
@@ -56,8 +56,7 @@ fi
 
 mkdir -p $BUILD_DIR
 cd $BUILD_DIR
-cmake \
-  -DCMAKE_TOOLCHAIN_FILE=../cmake/toolchains/emscripten.cmake \
+emcmake cmake \
```
Comment (ekpyron) [inline scripts/ci/build_emscripten.sh:59]: This is the canonical way to build cmake projects with emscripten. No idea why we actually ever bothered with the toolchain file mess...

File: cmake/EthCompilerSettings.cmake (hunk 2 above)
Comment (axic) [reply cmake/EthCompilerSettings.cmake:154]: That has been removed in 2.x.

**Comment by ekpyron** [comment]
All appears to be good. We could have a closer look at the CI artifact and check for performance changes, etc. - but I think this should generally be fine. CI seems to be faster or slower depending on the run, so I'd generally not expect much difference. Size is slightly larger, but didn't change significantly.

**Comment by ekpyron** [comment]
For the record: I did load it into remix in firefox and chrome and find any issues, but also didn't test exhaustively.

**Comment by ekpyron** [comment]
Hm... bytecode comparison complaining is actually rather concerning, though.

**Comment by ekpyron** [comment]
It appears to "only" be hashing in the metadata, though...
In particular the ipfs hash mismatches, because the swarm source hash in the metadata mismatches from what I can see...

**Comment by ekpyron** [comment]
Interestingly, both the keccak and the swarm hashes match, though, so it actually has to be the implementation of the swarm hash producing different results for the same input...
Generally I could imagine some 32-bit vs 64-bit weirdness, but that should have been noticable before the update as well... and it doesn't use boost from what I see... so it's actually rather weird...

**Comment by ekpyron** [comment]
The last commit should fix it.
For the record: quote https://en.cppreference.com/w/c/language/operator_arithmetic#Shift_operators 
"The behavior is undefined if rhs is negative or is greater or equal the number of bits in the promoted lhs."

Since wasm builds are 32-bit, size_t is 32-bit, so half of the shifts were undefined. Apparently old emscripten still used the "usual" shift semantics, i.e. all of this is zero, but new emscripten (respectively the new default optimization settings new emscripten uses) changes that.

**Comment by ekpyron** [comment]
Yep, all passes now.

**Review by chriseth — APPROVED** [review]

**Pull request merged**